### PR TITLE
 operators: Clean up and document lifecycle

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
1. Implement Close() method: Not having a Close() method was confusing
and led to situations when some resources could be leaked if there was a
failure in the operator.

2. Remove Init() and move that logic intro the corresponding Instantiate
method.

### CODE REVIEW

**Review by flyth — COMMENTED** [review]
Some initial feedback on the changes, will continue checking the implementation.

File: docs/devel/operator/lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+operator and what is their use case.
+
+## Priority
+
+Operators are initialized (PreStart and Start) according to their priority and
+they are closed (Stop, PreStop and Close) in the reverse order.
+
+## Lifecycle
+
+### Init
```
Comment (flyth) [inline docs/devel/operator/lifecycle.md:18]: This name clashes with `Init` from `DataOperator` - can we use something that makes it easily distinguishable whether it is an operator or an operator instance?

File: pkg/gadget-context/run.go [hunk 2]
```diff
@@ -18,12 +18,14 @@ import (
 	"fmt"
 	"sort"
 
+	"github.com/hashicorp/go-multierror"
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadget-context/run.go:21]: Opened #3841 to clean it repo wide in the meanwhile.

File: pkg/gadget-context/run.go [hunk 3]
```diff
@@ -88,79 +90,115 @@ func (c *GadgetContext) initAndPrepareOperators(paramValues api.ParamValues) ([]
...
+	if err := c.start(); err != nil {
 		return fmt.Errorf("starting operators: %w", err)
 	}
 
 	c.Logger().Debugf("running...")
 
 	WaitForTimeoutOrDone(c)
-	c.stop(dataOperatorInstances)
+	if err := c.stop(); err != nil {
+		return fmt.Errorf("stopping operators: %w", err)
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadget-context/run.go:197]: I'm finding all of this a little bit complicated as the logic to handle a failure and "reverting the chain" is not that easy to implement. What about a simplified model:
- Hooks are called in the following order: Init() -> PreStart() -> Stop() -> PostStop( ) -> Close()
- If any of the hooks fails, the next ones are skipped 
- Close() is always called
- Some hooks can be skipped even if the previous ones worked fine (in case another operator failed)

[the remark this answers is not in the document]
Comment (flyth) [reply pkg/gadget-context/run.go:197]: Hmm. Tricky.

> If any of the hooks fails, the next ones are skipped
Which ones do you mean? All next calls for the failing operator (e.g. Start() fails for one operator, don't call any other lifecycle hook for THAT operator), all same calls to the other operators (Start() fails for one, don't call the other operators' Start()), or both?

I think that after a successful initialization, we should call all lifecycle hooks for all initialized operators (=> most simple solution, IMHO). If the operator needs to keep state of a success/fail in one of the hooks, it should do it by itself. Maybe we should remove the `error` result of the hooks to make that even clearer.

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadget-context/run.go:197]: > Which ones do you mean? All next calls for the failing operator (e.g. Start() fails for one operator, don't call any other lifecycle hook for THAT operator), all same calls to the other operators (Start() fails for one, don't call the other operators' Start()), or both?

The ones for the same operator.

> I think that after a successful initialization, we should call all lifecycle hooks for all initialized operators (=> most simple solution, IMHO). If the operator needs to keep state of a success/fail in one of the hooks, it should do it by itself. Maybe we should remove the error result of the hooks to make that even clearer.

What's successful initialization? After Start()?

[the remark this answers is not in the document]
Comment (flyth) [reply pkg/gadget-context/run.go:197]: > What's successful initialization? After Start()?

InstantiateDataOperator / InstantiateImageOperator

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadget-context/run.go]: So, if preStart() fails, should we continue and call start(), stop(), etc? It seems confusing to me. I think it'll simplify the logic on the gadget context, but it'll make it harder to implement operators as they'd need to keep this state.

[the remark this answers is not in the document]
Comment (flyth) [reply pkg/gadget-context/run.go]: Yeah - I mean technically, if we remove the error, they can't really fail anymore, hehe. From that you then also directly know that there cannot be a fail state stored in the caller which could affect other lifecycle hooks.

Multiple thoughts: if a lifecycle hook (currently) fails, we just log it, right? So the error returned is just a convenient way of logging/exiting right now. If we skip certain hooks based on other hooks' outcomes, it adds this heavier logic that needs to be documented and understood properly - it also feels like it's calling for exceptions ("but I still need PostStop() to do X/Y if PostStart() fails but Start() succeeded"). Also, if for example you subscribe to events in PreStart and then fail, the callback will still be called despite failing - the operator has to take care of that as well. So in order to keep it consistent, I think it's fair to only allow failing in the init part (which will cancel the entire gadget run) and then handover responsibility to the operator and just blindly call all hooks the operator has.

So, bottom line ground rules: Close() has to tear down everything the operator created over its lifetime. And all hooks should carefully check whether things have been initialized correctly (if any) from any lifecycle hook before using them.

The added complexity in operators will probably only hit the more complex operators (ebpf, wams) anyway, but I'll check.

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadget-context/run.go]: > if a lifecycle hook (currently) fails, we just log it, right? So the error returned is just a convenient way of logging/exiting right now

It depends on the hook I think. IMO hooks like `Start()` should return an error to stop execution immediately if something goes wrong. Otherwise we'll be running a gadget with a "weird behaviour" (with only some operators running), etc.

> If we skip certain hooks based on other hooks' outcomes, it adds this heavier logic that needs to be documented and understood properly

It seems to me the logic to handle this on each operator (checking that everything was initialized properly) will be almost as heavier as the logic implemented in the gadget context with the additional problem that we'll need to implement it multiple times on different operators.

[the remark this answers is not in the document]
Comment (flyth) [reply pkg/gadget-context/run.go]: > > if a lifecycle hook (currently) fails, we just log it, right? So the error returned is just a convenient way of logging/exiting right now
> 
> It depends on the hook I think. IMO hooks like `Start()` should return an error to stop execution immediately if something goes wrong. Otherwise we'll be running a gadget with a "weird behaviour" (with only some operators running), etc.

Another option could be to give operators the possibility to tear down a gadget by canceling it using the context.

> > If we skip certain hooks based on other hooks' outcomes, it adds this heavier logic that needs to be documented and understood properly
> 
> It seems to me the logic to handle this on each operator (checking that everything was initialized properly) will be almost as heavier as the logic implemented in the gadget context with the additional problem that we'll need to implement it multiple times on different operators.

It really depends on the operator, I guess. Most of them are simple enough and won't need the logic.

However, I think both solutions have their pro/cons and there's no silver bullet. If you want to go with the other option, I'm fine with that.

File: pkg/gadget-context/run.go [hunk 4]
```diff
@@ -88,79 +90,115 @@ func (c *GadgetContext) initAndPrepareOperators(paramValues api.ParamValues) ([]
...
+	started := []operators.DataOperatorInstance{}
+
+	for _, opInst := range c.dataOperatorInstances {
 		c.Logger().Debugf("starting op %q", opInst.Name())
 		err := opInst.Start(c)
 		if err != nil {
-			c.cancel()
+			// Stop all operators that were started
+			for _, startedOp := range started {
+				startedOp.Stop(c)
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadget-context/run.go:118]: Makes sense, added PostStop() here.

File: pkg/gadget-context/run.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadget-context/run.go:200]: I changed my mind a bit (after trying to rework it multiple times during few hours).

Current approach is:
- if prestart or start fail, the chain is cancelled and an error is returned to the user 
- stop() and poststop() are guarantee to be called is start succeeds 
- close() is always called 

There is still a corner case: it's possible to call poststop() on an operator that returned error on stop(). I think we can live with it.

File: docs/devel/operator/lifecycle.md (hunk 1 above)
Comment (mauriciovasquezbernal) [reply docs/devel/operator/lifecycle.md]: This was removed.

**Comment by mauriciovasquezbernal** [comment]
I reworked it a bit. Please review it with care folks, I'm not completely sure it doesn't break things.

**Comment by mauriciovasquezbernal** [comment]
Added some tests that will hopefully make the intended behavior more clear.

**Review by flyth — COMMENTED** [review]
Just submitting some comments from earlier for now.

**Review by flyth — COMMENTED** [review]
Do you know whether this plays nicely with #4396?

I will continue reviewing the rest.

File: docs/devel/operator/lifecycle.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+they are closed (Stop, PreStop and Close) in the reverse order.
+
+## Lifecycle
+
+### Instantiate
+
+Instances (`DataOperatorInstance` and `ImageOperatorInstance`) are created by
+`InstantiateDataOperator()` and `InstantiateImageOperator()` respectively. Those
+functions should return a new initalized instance of the operator. They can also
+return `nil` is the operator should be skipped. Data sources should be created
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply docs/devel/operator/lifecycle.md:23]: Accepting the suggestion without the `,`

**Comment by mauriciovasquezbernal** [comment]
> Do you know whether this plays nicely with https://github.com/inspektor-gadget/inspektor-gadget/pull/4396?

It seems so, I rebased that branch on top of it (https://github.com/inspektor-gadget/inspektor-gadget/commit/254c6269acc8b539903dcf008a6547814faf3d7d) and everything seems to be working fine.

**Review by alban — COMMENTED** [review]
# :warning: **Performance Alert** :warning:

Possible performance regression was detected for benchmark **'Gadget benchmarks'**.
Benchmark result of this commit is worse than the previous benchmark result exceeding threshold `2`.

| Benchmark suite | Current: 8e3af9c8cb4e4189f70df241aeeb99e6c5f3ea3f | Previous: b4bbe1ccd4d3d91cb3a6b1605963616870d1cf23 | Ratio |
|-|-|-|-|
| `BenchmarkFormatter` | `584.9` ns/op | `244.3` ns/op | `2.39` |

This comment was automatically generated by [workflow](https://github.com/inspektor-gadget/inspektor-gadget/actions?query=workflow%3AInspektor%20Gadget%20CI) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).

CC: @alban

File: pkg/operators/ebpf/ebpf.go [hunk 6]
```diff
@@ -149,14 +149,10 @@ func (o *ebpfOperator) InstantiateImageOperator(
 
 	err = newInstance.init(gadgetCtx)
 	if err != nil {
+		newInstance.Close(gadgetCtx)
```
Comment (flyth) [inline pkg/operators/ebpf/ebpf.go:154]: Maybe add a comment on why this needs to be called here?

File: pkg/operators/ebpf/ebpf.go [hunk 7]
```diff
@@ -345,6 +341,89 @@ func (i *ebpfInstance) init(gadgetCtx operators.GadgetContext) error {
 		return fmt.Errorf("initializing formatters: %w", err)
 	}
 
+	if err := i.evaluateMapParams(i.paramValues); err != nil {
+		return fmt.Errorf("evaluating map params: %w", err)
+	}
+
+	for ds, formatters := range i.formatters {
```
Comment (flyth) [inline pkg/operators/ebpf/ebpf.go]: This should happen on PreStart(), right?

File: pkg/operators/ebpf/ebpf.go [hunk 8]
```diff
@@ -345,6 +341,89 @@ func (i *ebpfInstance) init(gadgetCtx operators.GadgetContext) error {
...
+		for _, formatter := range formatters {
+			formatter := formatter
```
Comment (flyth) [inline pkg/operators/ebpf/ebpf.go]: Not necessary anymore since Go 1.22.

File: pkg/operators/localmanager/localmanager.go [hunk 9]
```diff
@@ -643,10 +643,10 @@ func (l *localManagerTraceWrapper) Start(gadgetCtx operators.GadgetContext) erro
 }
 
 func (l *localManagerTraceWrapper) Stop(gadgetCtx operators.GadgetContext) error {
-	if l.containersPublisher != nil {
-		l.containersPublisher.Unsubscribe()
-	}
+	return nil
```
Comment (flyth) [inline pkg/operators/localmanager/localmanager.go:650]: Is the previous Unsubscribe() not needed anymore?

File: pkg/operators/oci-handler/oci.go [hunk 10]
```diff
@@ -503,25 +505,41 @@ func (o *OciHandlerInstance) PreStop(gadgetCtx operators.GadgetContext) error {
 }
 
 func (o *OciHandlerInstance) Stop(gadgetCtx operators.GadgetContext) error {
+	var errs []error
+
 	for _, opInst := range o.imageOperatorInstances {
 		err := opInst.Stop(o.gadgetCtx)
 		if err != nil {
-			o.gadgetCtx.Logger().Errorf("stopping operator %q: %v", opInst.Name(), err)
+			errs = append(errs, fmt.Errorf("stopping operator %q: %w", opInst.Name(), err))
```
Comment (flyth) [inline pkg/operators/oci-handler/oci.go:514]: No need to handle this in an array, btw. Can be simplified by having `(err error)` in the signature and then just doing `err = errors.Join(err, fmt.Errorf(...))` - `errors.Join()` will return nil, if there's no error.

**Review by flyth — COMMENTED** [review]
Changes look good, overall. I'm a bit worried about the ebpf operator, it's hard to follow all the things we're doing in there - but let's revisit that after merging this PR.

File: pkg/operators/ebpf/ebpf.go (hunk 7 above)
Comment (mauriciovasquezbernal) [reply pkg/operators/ebpf/ebpf.go]: Yeah, that's even, even if the previous approach was working as those datasources are created just before by the same ebpf operator.

File: pkg/operators/localmanager/localmanager.go (hunk 9 above)
Comment (mauriciovasquezbernal) [reply pkg/operators/localmanager/localmanager.go:650]: I think it's still needed! Good catch!

File: pkg/operators/oci-handler/oci.go (hunk 10 above)
Comment (mauriciovasquezbernal) [reply pkg/operators/oci-handler/oci.go:514]: But that `err` on the signature will collide with the one on L511. I'll keep this way for now.

**Review by flyth — APPROVED** [review]
Thanks for cleaning this up!

**Comment by mauriciovasquezbernal** [comment]
Thanks a lot for the reviews, I hope it helps in the long term.
